src/replicate_stage.rs

```diff
@@ -1,57 +1,133 @@
+use packet::BlobRecycler;
 use result::{Error, Result};
 use service::Service;
+use signature::KeyPair;
+use std::collections::VecDeque;
+use std::net::UdpSocket;
+use std::sync::mpsc::channel;
 use std::sync::mpsc::RecvTimeoutError;
-use std::sync::Arc;
+use std::sync::{Arc, RwLock};
 use std::thread::{self, Builder, JoinHandle};
 use std::time::Duration;
-use streamer::BlobReceiver;
+use streamer::{responder, BlobReceiver, BlobSender};
+use timing;
+use transaction::Transaction;
+use voting::entries_to_votes;
 
 pub struct ReplicateStage {
-    thread_hdl: JoinHandle<()>,
+    thread_hdls: Vec<JoinHandle<()>>,
 }
 
+const VOTE_TIMEOUT_MS: u64 = 1000;
+
 impl ReplicateStage {
     /// Process entry blobs, already in order
-    fn replicate_requests(bank: &Arc<Bank>, blob_receiver: &BlobReceiver) -> Result<()> {
+    fn replicate_requests(
+        keypair: &Arc<KeyPair>,
+        bank: &Arc<Bank>,
+        crdt: &Arc<RwLock<Crdt>>,
+        blob_recycler: &BlobRecycler,
+        window_receiver: &BlobReceiver,
+        vote_blob_sender: &BlobSender,
+        last_vote: &mut u64,
+    ) -> Result<()> {
         let timer = Duration::new(1, 0);
-        let blobs = blob_receiver.recv_timeout(timer)?;
+        //coalesce all the available blobs into a single vote
+        let mut blobs = window_receiver.recv_timeout(timer)?;
+        while let Ok(mut more) = window_receiver.try_recv() {
+            blobs.append(&mut more);
+        }
         let blobs_len = blobs.len();
-        let entries = ledger::reconstruct_entries_from_blobs(blobs)?;
+        let entries = ledger::reconstruct_entries_from_blobs(blobs.clone())?;
+        let votes = entries_to_votes(&entries);
         let res = bank.process_entries(entries);
         if res.is_err() {
             error!("process_entries {} {:?}", blobs_len, res);
         }
-        res?;
+        let now = timing::timestamp();
+        if now - *last_vote > VOTE_TIMEOUT_MS {
+            let height = res?;
+            let last_id = bank.last_id();
+            let shared_blob = blob_recycler.allocate();
+            let (vote, addr) = {
+                let mut wcrdt = crdt.write().unwrap();
+                wcrdt.insert_votes(votes);
+                //TODO: doesn't seem like there is a synchronous call to get height and id
+                info!("replicate_stage {} {:?}", height, &last_id[..8]);
+                let (vote, addr) = wcrdt.new_vote(height, last_id)?;
+                (vote, addr)
+            };
+            {
+                let mut blob = shared_blob.write().unwrap();
+                let tx = Transaction::new_vote(&keypair, vote, last_id, 0);
+                let bytes = serialize(&tx)?;
+                let len = bytes.len();
+                blob.data[..len].copy_from_slice(&bytes);
+                blob.meta.set_addr(&addr);
+                blob.meta.size = len;
+            }
+            *last_vote = now;
+            vote_blob_sender.send(VecDeque::from(vec![shared_blob]))?;
```

Comment:
I'm curious, why are we sending a `VecDeque` on a channel?

Reply:
Its a system call, so normally we would be sending a batch of things to do all in one go. just in this case there is only 1 item.

Reply:
But why a `VecDeque` and not a `Vec`?

Reply:
i am not sure :)

Reply:
added  #572